pkg/resources/stream_on_table.go

```diff
@@ -0,0 +1,307 @@
+	BeforeAttributeName: schemas.BeforeSchema,
+	ShowOutputAttributeName: {
+		Type:        schema.TypeList,
+		Computed:    true,
+		Description: "Outputs the result of `SHOW STREAMS` for the given stream.",
+		Elem: &schema.Resource{
+			Schema: schemas.ShowStreamSchema,
+		},
+	},
+	DescribeOutputAttributeName: {
+		Type:        schema.TypeList,
+		Computed:    true,
+		Description: "Outputs the result of `DESCRIBE STREAM` for the given stream.",
+		Elem: &schema.Resource{
+			Schema: schemas.DescribeStreamSchema,
+		},
+	},
+	FullyQualifiedNameAttributeName: schemas.FullyQualifiedNameSchema,
+}
+
+func StreamOnTable() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: CreateStreamOnTable(false),
+		ReadContext:   ReadStreamOnTable(true),
+		UpdateContext: UpdateStreamOnTable,
+		DeleteContext: DeleteStreamContext,
+		Description:   "Resource used to manage streams on tables. For more information, check [stream documentation](https://docs.snowflake.com/en/sql-reference/sql/create-stream).",
+
+		CustomizeDiff: customdiff.All(
+			ComputedIfAnyAttributeChanged(streamOnTableSchema, ShowOutputAttributeName, "table", "append_only", "comment"),
+			ComputedIfAnyAttributeChanged(streamOnTableSchema, DescribeOutputAttributeName, "table", "append_only", "comment"),
+		),
+
+		Schema: streamOnTableSchema,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: ImportStreamOnTable,
+		},
+	}
+}
+
+func ImportStreamOnTable(ctx context.Context, d *schema.ResourceData, meta any) ([]*schema.ResourceData, error) {
+	log.Printf("[DEBUG] Starting stream import")
+	client := meta.(*provider.Context).Client
+	id, err := sdk.ParseSchemaObjectIdentifier(d.Id())
+	if err != nil {
+		return nil, err
+	}
+
+	v, err := client.Streams.ShowByID(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+	if err := d.Set("name", id.Name()); err != nil {
+		return nil, err
+	}
+	if err := d.Set("database", id.DatabaseName()); err != nil {
+		return nil, err
+	}
+	if err := d.Set("schema", id.SchemaName()); err != nil {
+		return nil, err
+	}
+	if err := d.Set("append_only", booleanStringFromBool(*v.Mode == sdk.StreamModeAppendOnly)); err != nil {
+		return nil, err
+	}
+	return []*schema.ResourceData{d}, nil
+}
+
+func CreateStreamOnTable(orReplace bool) schema.CreateContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, meta any) diag.Diagnostics {
+		client := meta.(*provider.Context).Client
+		databaseName := d.Get("database").(string)
+		schemaName := d.Get("schema").(string)
+		name := d.Get("name").(string)
+		id := sdk.NewSchemaObjectIdentifier(databaseName, schemaName, name)
+
+		tableIdRaw := d.Get("table").(string)
+		tableId, err := sdk.ParseSchemaObjectIdentifier(tableIdRaw)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		req := sdk.NewCreateOnTableStreamRequest(id, tableId)
+		if orReplace {
+			req.WithOrReplace(true)
+		}
+
+		if v := d.Get("copy_grants"); v.(bool) {
+			req.WithCopyGrants(true).WithOrReplace(true)
```

Comment:
is OrReplace needed here? It allows for possible overriding of already existing stream, without it the task will use OrReplace only when modified attribute cannot be altered.

Reply:
It's needed because otherwise when you have COPY GRANTS without OR REPLACE, Snowflake fails.

Reply:
You are right about the need for orReplace needed with copy grants but I am with @sfc-gh-jcieslak here: we should steer the copy grants logic, and use it only when we need to replace the stream and not by default in create

Reply:
We don't use OR REPLACE by default. It's used when the stream is already in the state, and detect changes on fields that can't be updated, because ForceNew drops the grants. In the second case, when it's not in the state, and the user wants COPY GRANTS to preserve old grants.

Reply:
I disagree with the second case. IMO a new create should be done without copy grants and without or replace. Thge assumption should be, that if we are creating a new object, the copy grants do not matter. Copy grants should be used only in the first scenario you mentioned (so the object is already managed and we discover changes).

Reply:
Changed again.